### PR TITLE
Add buffers and tabs component

### PR DESCRIPTION
**Pull request by kdheepak**

### CODE REVIEW

**Comment by shadmansaleh** [comment]
Let me know when you're ready for review .

Do you think we should have additional bufferline component ?

Also I think we should give an option tabline shows name / no for tabs.

For transitional separators,
Insert `%s{sep}` to display sep as left separator and `%S{sep}` to display sep as right separator . Also inspect the options table there are config info and other info like which section the component is on IRRC.

**Comment by kdheepak** [comment]
> Do you think we should have additional bufferline component ?

Yes, I think we should add a bufferline component too. I'll add this first. And once I've gotten a hang of it, I'll update the PR with buffers too.

> Also I think we should give an option tabline shows name / no for tabs.

What should the name for tabs be? User provided? That feature is available in `tabline.nvim`, and it is done by storing the name for each tab in a global variable.

**Comment by kdheepak** [comment]
Do you know how I can get the active and inactive highlights? The value would change depending on which section the user chooses to have this component in, right?

**Comment by shadmansaleh** [comment]
> What should the name for tabs be? User provided? That feature is available in tabline.nvim

I was thinking more like the default tabline . It shows filename of last open buffer in the tab I think .

**Comment by shadmansaleh** [comment]
> Do you know how I can get the active and inactive highlights?

update_status gets a second parameter is_active a boolean value that lets you know whether that component is being evaluated for inactive or active status .

What do you actually mean by get ? like extract_hl ?

**Comment by kdheepak** [comment]
I imagine we want something like this, right?

Screen Shot 2021-09-10 at 11 28 41 AM

We want a `buffers` component and a `tabs` component. And I expect users will configure it like so:

```lua
require("lualine").setup({
  ...
  tabline = {
    lualine_a = {"buffers"},
    lualine_b = {},
    lualine_c = {},
    lualine_x = {},
    lualine_y = {},
    lualine_z = {"tabs"},
  },
  ...
```

But someone might want to use `{"buffers"}` in `lualine_c` for example. And the highlight for the "inactive" and "active" buffers will be different based on which section it is in. I was just wondering if you had an easier way to do it or any suggestions for how to go about doing this.

**Comment by shadmansaleh** [comment]
I think for active buffer/tab you can directly  use `lualine_a_normal` hl group.  for the inactive hl get the defualt hl for the section by calling `highlight.format_highlight` like [here](https://github.com/shadmansaleh/lualine.nvim/blob/1f4df948c4a95e5ed276a29f6916e46735ff411a/lua/lualine/components/filetype.lua#L39) . It'll give what highlight should be on that section . You can use that to make the component blend with the section .

**Comment by kdheepak** [comment]
I was expecting something like this to work:

```lua
%#lualine_b_normal#%1@LualineSwitchTab@1 %T%#lualine_b_normal#%S{}%#lualine_a_normal#%2@LualineSwitchTab@ 2 %T%#lualine_a_normal#"
```

Am I not using the separators correctly? This is rendering like this:


Screen Shot 2021-09-17 at 9 56 30 PM

**Comment by shadmansaleh** [comment]
You should use %s (Just try with %s when %S doesn't work and try with %S when %s doesn't work. transitional separators are a pain :P). Also you need to use section_separators
try
```
%#lualine_b_normal#%1@LualineSwitchTab@1 %T%#lualine_b_normal#%s{}%#lualine_a_normal#%2@LualineSwitchTab@ 2 %T%#lualine_a_normal#
```

**Comment by kdheepak** [comment]
Thoughts about the current implementation?


https://user-images.githubusercontent.com/1813121/133915320-729d9213-2b34-473c-85f0-261aba4c6323.mov

If this looks good, I'll port the same to get buffers working too.

**Comment by kdheepak** [comment]
I've added a `buffers` component too. However, with the following configuration:

```lua
  tabline = {
    lualine_a = { "buffers" },
    lualine_b = {},
    lualine_c = {},
    lualine_x = {},
    lualine_y = {},
    lualine_z = { "tabs" },
  },
```

I get the following tabline:
Screen Shot 2021-09-18 at 11 03 17 PM

Screen Shot 2021-09-18 at 11 03 31 PM

Screen Shot 2021-09-18 at 11 03 46 PM

If I use the following configuration however:

```
  tabline = {
    lualine_a = { "buffers" },
    lualine_b = { "mode" },
    lualine_c = {},
    lualine_x = {},
    lualine_y = { "mode" },
    lualine_z = { "tabs" },
  },
```

I get this:

Screen Shot 2021-09-18 at 11 06 04 PM

It seems like the empty sections always follow whatever is the last highlight in the previous section.

**Comment by shadmansaleh** [comment]
rebase on master and see if https://github.com/shadmansaleh/lualine.nvim/commit/7147989f5f216852dabd0c751df7249e420385b6 solves the issue

**Comment by kdheepak** [comment]
Thanks for the suggestion! It does seem to solve it. However, now the `statusline` is showing some odd colors:

Screen Shot 2021-09-19 at 12 02 55 AM

**Comment by shadmansaleh** [comment]
What's your statusline config ?

Also do another rebase I've pushed another commit.
See if that fixes the statusline issue

signature of `highlight.format_highlight` has changed a bit now `is_focused` is second and optional argument.

**Comment by shadmansaleh** [comment]
I've rebased it on master and fixed the format_highlight issues that occured . Check if it's working correctly now.

**Comment by kdheepak** [comment]
That seems to work!

**Comment by kdheepak** [comment]
I think this is ready for you to look at now.

**Comment by shadmansaleh** [comment]
Ok I'll check it out once I get time . Maybe sort out the lint issues :D

File: lua/lualine/components/tabs.lua [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+  self.__index = self
+  newObj = setmetatable(newObj, self)
+  return newObj
+end
+
+function Tab:render()
+  local line = ''
+  local name = self.name .. ' '
+  if self.last then
+    name = ' ' .. name
```
Comment (shadmansaleh) [inline lua/lualine/components/tabs.lua:28]: I'm not sure if I understand this
is this what you want
- theres a post space always (even in last tab)
- there's a pre space on every tab except 1st tab (1st tab has pre space when it's only space)

should there be a post space on last tab ?

**Comment by shadmansaleh** [comment]
@kdheepak I've made some changes to tabs in https://github.com/kdheepak/lualine.nvim/pull/42/commits/95925cd794061a8cde220c6fcffc14bd0b78905c let me know what you think . The functionality should stay the same.

**Comment by shadmansaleh** [comment]
I think it'll be nicer if we use same sections active and inactive hl to indicate active and inactive state in buffers/tabs. I've made it like that.

Take a look at the changes I made to buffers & tabs and let me know if you have any objections.

I think one major feature we are missing is providing an option to show last used buffers bufname in tabs like builtin tabline .

One more thing I noticed is buffers like fzf don't really appear on buffers . I checked the fzf buffer was marked unlisted that's why it's getting filtered out I think . But on tabline.nvim it is shown any idea why ?

**Comment by kdheepak** [comment]
> One more thing I noticed is buffers like fzf don't really appear on buffers . I checked the fzf buffer was marked unlisted that's why it's getting filtered out I think . But on tabline.nvim it is shown any idea why ?

That's strange. I'm looking into it.

**Comment by kdheepak** [comment]
> One more thing I noticed is buffers like fzf don't really appear on buffers . I checked the fzf buffer was marked unlisted that's why it's getting filtered out I think . But on tabline.nvim it is shown any idea why ?

I think it is because of when `update_status` is called. It appears to be different for a `Component` vs a `function`

**Comment by shadmansaleh** [comment]
> I think it is because of when update_status is called. It appears to be different for a Component vs a function

That doesn't make sense . functions are just like components . and the function is called from update_status how will it be different

https://github.com/shadmansaleh/lualine.nvim/blob/28d89472e3326f81d3d7e4ff0054723eeade9e3d/lua/lualine/components/special/function_component.lua#L3-L17

**Comment by kdheepak** [comment]
> I think it'll be nicer if we use same sections active and inactive hl to indicate active and inactive state in buffers/tabs. I've made it like that.

I agree! One of my first changes was to make it like this. But for some reason I went back on it. I think your suggestion and your changes fits `lualine`'s styling a lot better.

> Take a look at the changes I made to buffers & tabs and let me know if you have any objections.

Screen Shot 2021-09-19 at 8 30 45 AM
Screen Shot 2021-09-19 at 8 31 01 AM

1. `section_c` ends with a `|` instead of `>`. Is that possible to fix?
2. In `section_b` there's no way to distinguish between the current buffer and other open buffers

**Comment by shadmansaleh** [comment]
> section_c ends with a | instead of >. Is that possible to fix?

That's section boundary that's why component separators aren't applied . Is that a serous issue?

> In section_b there's no way to distinguish between the current buffer and other open buffers

I think it's because of the theme you're using . It has same color for active and inactive status on section_b

**Comment by shadmansaleh** [comment]
https://github.com/kdheepak/lualine.nvim/pull/42/commits/5b2592f76b2c30e2bb2d7afbe768e664e31cd389

this seems to make color changes in active&inactive buffer look weird. What's the issue with section separators ?

**Comment by shadmansaleh** [comment]
> >In section_b there's no way to distinguish between the current buffer and other open buffers
>
>I think it's because of the theme you're using . It has same color for active and inactive status on section_b

I think we should expose active and inactive color as options (like diff/diagnostics components) . That'll cover cases where themes do that .

**Comment by kdheepak** [comment]
> this seems to make color changes in active&inactive buffer look weird. What's the issue with section separators ?

Screen Shot 2021-09-19 at 9 06 54 AM

I think with your change, we don't want to use section separators anywhere, right? We want to use component separators.

**Comment by shadmansaleh** [comment]
> I think with your change, we don't want to use section separators anywhere, right?

I'm in favor of using section_separators .

Were they causing any issue with my changes ? I'm not sure why you removed them.

**Comment by kdheepak** [comment]
See the previous set of screenshots. There’s no separator between the current buffer name and the next buffer name. I’m on my phone again now, otherwise I’d have shared more screenshots.

**Comment by shadmansaleh** [comment]
I think separator is not drawn because both active and inactive highlight groups have same bg color . It does work when they are different .

Screenshot

I think if we make active and inactive colors customizable we can call it a non issue ?

**Comment by kdheepak** [comment]
Awesome! That sounds good to me!

**Comment by kdheepak** [comment]
> That doesn't make sense . functions are just like any other component. and the function is called from update_status how will it be different

I think I know why `tabline` is displaying it differently. I'll make a change and push it in a bit.

**Comment by kdheepak** [comment]
@shadmansaleh Can you try it now to see how it behaves with fzf?

**Comment by kdheepak** [comment]
> I think one major feature we are missing is providing an option to show last used buffers bufname in tabs like builtin tabline .

I've added this in the most recent commit.

File: lua/lualine/components/tabs.lua (hunk 1 above)
Comment (kdheepak) [reply lua/lualine/components/tabs.lua]: This is cleaned up now and resolved.

**Comment by shadmansaleh** [comment]
Great it works now .

Can you make it so buffers/tabs doesn't take over entire tabline by default . like in tabline.nvim where `...` is shown when they get too big . I really like that :)

**Comment by kdheepak** [comment]
Done! Can you try it out?

Screen Shot 2021-09-19 at 9 21 27 PM

**Comment by shadmansaleh** [comment]
I think highlights are broken with ellipsis . Thry should be displayed as inactive not active .

Also long concats are bad for performance as lua needs to allocate a new string from every concat .Even lua handbook warns against it . That's why I swiched to using `table.concat` .

**Comment by kdheepak** [comment]
Ah, thanks for letting me know. I'll update based on your comments.

**Comment by shadmansaleh** [comment]
issue template fix should go on separate pr

**Comment by kdheepak** [comment]
Got it. I removed the commit from this PR.

**Comment by kdheepak** [comment]
I think this should be ready for final review / merging now.

**Comment by shadmansaleh** [comment]
> I think this should be ready for final review / merging now.

Yeah it pretty complete . I'll do a few more tweeks and merge it I think

**Comment by kdheepak** [comment]
Great! Thanks! Let me know if you need anything else from me.

**Comment by shadmansaleh** [comment]
I think we can merge this if you're ok with it.

**Comment by kdheepak** [comment]
Your changes look good to me! Thanks for making all of them!

**Comment by shadmansaleh** [comment]
Thanks for this . Really appreciate it :)

**Comment by kdheepak** [comment]
Thanks for all your work on this PR and this plugin! You are awesome!

**Comment by Hubro** [comment]
Not sure if I'm doing something wrong here, but with this PR I can no longer tell which tab is active. Here's how "tabs" look in various themes:

image

image

image

image

image

(The active tab is "2" in all these screenshots)

When using `require("lualine").tabline_tabs`:

image

image

etc.

My config is just `lualine_x = { "tabs" }`.

**Comment by kdheepak** [comment]
Colorschemes need to add support for different colors for active and inactive highlights, and it’ll work. I would recommend opening an issue on the repo of the corresponding colorscheme. 

You can also manually configure the colors in your lualine config, see README for how to do that.

**Comment by Hubro** [comment]
@kdheepak Why isn't this necessary with `tabline`?

**Comment by kdheepak** [comment]
Because tabline “hardcodes” the section separator and components separators and highlights appropriately. This is why tabline only works in section_z at the moment. The implementation in lualine however should work in any section. 

@shadmansaleh thoughts? Perhaps we could use a better default than instead of the current inactive highlight? At least till more colorschemes explicitly support this.

**Comment by shadmansaleh** [comment]
If theme is inadequate user can just specify the active or inactive colors . If it wasn't configurable I'd consider this an issue. Also tabs and buffers are just another component so it makes sense that they use colors from the section they are in by default.

**Pull request merged**